### ISSUE TITLE
tf.nn.moments produces NaNs with axes=[-1]

### ISSUE THREAD

**Issue by jozef-mokry** [tensorflow/tensorflow#8101]
Hi,

tf.nn.moments does not work with negative axes. Is this intentional? Will this be possible in future versions? Thanks.

```
>>> x = tf.placeholder(dtype=tf.float32)
>>> mean, var = tf.nn.moments(x, axes=[1])
>>> sess.run([mean, var], {x:np.arange(12).reshape((2,6)).astype(np.float32)})
[array([ 2.5,  8.5], dtype=float32), array([ 2.91666675,  2.91666675], dtype=float32)]
>>> mean, var = tf.nn.moments(x, axes=[-1])
>>> sess.run([mean, var], {x:np.arange(12).reshape((2,6)).astype(np.float32)})
[array([ nan,  nan], dtype=float32), array([ nan,  nan], dtype=float32)]
```

**Comment by prb12** [comment]
I don't see anything in the documentation which suggests that tf.nn.moments should accept axes <0.  To be clear, you would like axis -N to refer to dimension rank-N?

**Comment by jozef-mokry** [comment]
Yes, I was hoping -1 would refer to the last axis (and rank-N in general). I agree the documentation does not mention this. However for functions like reduce_sum, it is not mentioned explicitly either (except for the "Equivalent to np.sum" note) but they do support negative axes. 

It is quite tricky to remember which functions support negative axes and which don't and if you get it wrong you get NaNs. Maybe a simple 'assert all axes are non-negative' might prevent a lot of time debugging NaNs in the future.

**Comment by prb12** [comment]
@jozef-mokry Yes, it does seem reasonable that the axes argument should be validated in this case. Are you running on GPU or CPU?

**Comment by jozef-mokry** [comment]
I ran this on a GPU.

**Comment by yzhwang** [comment]
@drpngx Any update on this?

**Comment by yongtang** [comment]
I tried with the latest master and it seems the issue has been fixed:
```
ubuntu@ip-172-31-28-246:~$ python
Python 2.7.12 (default, Nov 19 2016, 06:48:10) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> import tensorflow as tf
>>> sess = tf.Session()
2017-09-01 19:55:56.968431: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:892] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2017-09-01 19:55:56.968903: I tensorflow/core/common_runtime/gpu/gpu_device.cc:965] Found device 0 with properties: 
name: Tesla K80 major: 3 minor: 7 memoryClockRate(GHz): 0.8235
pciBusID: 0000:00:1e.0
totalMemory: 11.17GiB freeMemory: 11.11GiB
2017-09-01 19:55:56.968928: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1055] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:00:1e.0, compute capability: 3.7)
>>> 
>>> x = tf.placeholder(dtype=tf.float32)
>>> mean, var = tf.nn.moments(x, axes=[1])
>>> sess.run([mean, var], {x:np.arange(12).reshape((2,6)).astype(np.float32)})
[array([ 2.5,  8.5], dtype=float32), array([ 2.91666675,  2.91666675], dtype=float32)]
>>> mean, var = tf.nn.moments(x, axes=[-1])
>>> sess.run([mean, var], {x:np.arange(12).reshape((2,6)).astype(np.float32)})
[array([ 2.5,  8.5], dtype=float32), array([ 2.91666675,  2.91666675], dtype=float32)]
>>> 
```

**Comment by drpngx** [comment]
Thank you for reporting back!
